src/main/java/be/seeseemelk/mockbukkit/MockUnsafeValues.java

```diff
@@ -296,32 +301,121 @@ public Entity deserializeEntity(byte[] data, World world, boolean preserveUUID)
+	 * @return a string of the structure "block.{@literal <namespace>}.{@literal <material>}" (e.g. "block.minecraft.stone"), or null if not a block.
+	 */
 	@Override
 	public String getBlockTranslationKey(Material material)
 	{
-		// TODO Auto-generated method stub
-		throw new UnimplementedOperationException();
+		if(!material.isBlock()) {
+			return null;
+		}
+		// edge cases: WHEAT and NETHER_WART are blocks, but still use the "item" prefix
+		if(material == Material.WHEAT || material == Material.NETHER_WART) {
+			return formatTranslatable("item", material);
+		}
+		return formatTranslatable("block", material);
 	}
 
+	/**
+	 * Gets the translation key for a {@link Material} that {@link Material#isItem() is an item}.
+	 * @param material the material to translate
+	 * @return a string of the structure "item.{@literal <namespace>}.{@literal <material>}" (e.g. "item.minecraft.carrot"), or null if not an item.
+	 */
 	@Override
 	public String getItemTranslationKey(Material material)
 	{
-		// TODO Auto-generated method stub
-		throw new UnimplementedOperationException();
+		if(!material.isItem()) {
+			return null;
+		}
+		String edgeCaseHandledTranslationKey = handleTranslateItemEdgeCases(material);
+		if(edgeCaseHandledTranslationKey != null) {
+			return edgeCaseHandledTranslationKey;
+		}
+		return formatTranslatable("item", material);
 	}
 
+	/**
+	 * Gets the translation key for an {@link EntityType}. Throws an error for custom entities.
+	 * @param type the entity to translate
+	 * @return a string of the structure "item.{@literal <namespace>}.{@literal <entity_type>}" (e.g. "entity.minecraft.pig").
+	 */
 	@Override
 	public String getTranslationKey(EntityType type)
 	{
-		// TODO Auto-generated method stub
-		throw new UnimplementedOperationException();
+		Preconditions.checkArgument(type.getName() != null, "Invalid name of EntityType %s for translation key", type);
+		// prevent translation from any custom entity type
+		Arrays.stream(EntityType.values())
+				.filter(entityType ->type.getName().equals(entityType.getName()))
+				.findFirst()
+				.orElseThrow();
+		return formatTranslatable("entity", type);
 	}
 
+	/**
+	 * Gets the translation key for a {@link ItemStack} that {@link Material#isItem() is an item}.
+	 * @param itemStack the itemstack to translate
+	 * @return a string of the structure "item.{@literal <namespace>}.{@literal <material>}" (e.g. "item.minecraft.carrot"), or null if not an item.
+	 */
 	@Override
 	public String getTranslationKey(ItemStack itemStack)
 	{
-		// TODO Auto-generated method stub
-		throw new UnimplementedOperationException();
+		if(itemStack.getType().isItem()) {
+			Material material = itemStack.getType();
+			if(!material.isItem()) {
+				return null;
+			}
+			String edgeCaseHandledTranslationKey = handleTranslateItemEdgeCases(material);
+			if(edgeCaseHandledTranslationKey != null) {
+				return edgeCaseHandledTranslationKey;
+			}
+			return formatTranslatable("item", material, true);
```

Comment:
Code duplication, use the getItemTranslationKey(Material) method

Reply:
I cant though. I need to somehow "mark" a Material that is "coming from an itemstack", because "empty effect" materials (POTION for example) have different translation keys when used from an itemstack vs. when used as a material.

Reply:
Ah, I missed the `true` mark there.